Improve CLI flag and help processing

pde opened the pull request:
* `letsencrypt --help $SUBCOMMAND` now works.
   Fixes #819
   Gets us closer to #787 
 * `letsencrypt -h $TOPIC` works
   Fixes #837 
 * subcommand arguments are now actually argument groups, so that all flags
   can be placed before or after subcommand verbs as the user wishes
   Fixes: #820  #442 

A limitation:

 * args like `--cert-path` were previously present for multiple verbs
   (auth/install/revoke) with separate docs; they are now in the "paths"
   topic.  That's fine, though it would be good to *also* list them when the
   user types `letsencrypt --help install`.

pde commented:
Ready for review, I think.

bmw commented on letsencrypt/cli.py:
```diff
@@ -739,62 +749,66 @@ def add_subparser(name, func):  # pylint: disable=missing-docstring
+    add_subparser("revoke", revoke)
+    helpful.add_group("revoke", description="Options for revocation of certs")
+    add_subparser("rollback", rollback)
+    helpful.add_group("rollback", description="Options for reverting config changes")
```
With everything moving to argument groups, what is the purpose of the subparsers? To control what function runs?

A bit of a nitpick, but I think like it would be clearer and less code to change `VERBS` to a `dict` that maps verbs to functions. You could still iterate over `VERBS` the same way and lookup the desired function in the bottom of `main`.

bmw commented:
This is certainly an improvement on the client (closing 5 issues) and is fine to merge. LGTM.

With that said, I think that the limitation you pointed out in your comment should certainly be addressed at some point and perhaps my nitpick as well. This also makes `pep8` unhappy, which certainly isn't a blocker, but is being followed by a lot of contributors.

pde commented:
@bmw I've implemented the VERBS table you suggested. Note that it will require slightly more elaborate mockery for our tests.

I continue to not be a big fan of PEP8 and pylint. They consume our time unhelpfully, and work often make the code less readable.  See the 28 pylint disables I just removed...

pde commented:
This passed integration tests when I ran them.  I'm not sure what's up with the Connection refused errors there...

bmw commented on acme/acme/challenges_test.py:
```diff
@@ -269,7 +269,7 @@ def setUp(self):
         }
 
-        # pylint: disable=invalid-name
+
```
Unnecessary newline. It looks like there are a few of these wherever `invalid-name` was the only option being temporarily disabled in `pylint`.

bmw commented on letsencrypt/cli.py:
```diff
@@ -739,62 +754,61 @@ def add_subparser(name, func):  # pylint: disable=missing-docstring
+    # attached, but they have annoying arg ordering constrains so we use
+    # groups instead: https://github.com/letsencrypt/letsencrypt/issues/820
+    for v, func in VERBS.items():
+        add_subparser(v, func)
```
I wasn't very clear with my original nitpick on this. What I meant by my suggestion was that maybe we should remove subparsers entirely as they seem like a heavyweight solution for mapping a string to a function (hence the `dict` idea).

kuba commented on .pylintrc:
```diff
@@ -38,7 +38,7 @@ load-plugins=linter_plugin
 # no Warning level messages displayed, use"--disable=all --enable=classes
 # --disable=W"
-disable=fixme,locally-disabled,abstract-class-not-used,bad-continuation,too-few-public-methods,no-self-use
+disable=fixme,locally-disabled,abstract-class-not-used,bad-continuation,too-few-public-methods,no-self-use,invalid-name
```
Please don't disable it, at least for the `acme` module together with removing all stanzas. And definitely not in an unrelated PR. Thanks in advance!

kuba commented on letsencrypt/cli.py:
```diff
@@ -723,10 +730,18 @@ def create_parser(plugins, args):
-VERBS = ["run", "auth", "install", "revoke", "rollback", "config_changes",
-         "plugins", "--help"]
-
+# there isn't an elegant way to autogenerate it in time. pylint: disable=bad-whitespace
```
rather than disabling the `bad-whitespace` (which in fact appies to the entire file, not only `VERBS`!), just fix the formatting not to indent; when we add new verb would you realign everything?

kuba commented on letsencrypt/cli.py:
```diff
@@ -739,62 +754,61 @@ def add_subparser(name, func):  # pylint: disable=missing-docstring
+    elif verb == "revoke":
+        add("paths", "--cert-path", type=read_file, required=True, help=cph)
+    else:
+        add("paths", "--cert-path", help=cph, required=(verb == "install"))
```
How does this new logic work with `--help all`? Would it display all versions of `--cert-path`?

pde replied:
It displays the third one.  They all have the same help string, perhaps that help string should note that the argument is required for `install` and `revoke`, though I'd prefer that we figure out how to make those verbs work without `--cert-path` :)

kuba commented on letsencrypt/tests/cli_test.py:
```diff
@@ -43,7 +43,11 @@ def _call(self, args):
+            from letsencrypt import cli
+            cli.VERBS["run"] = mock_run
+            self._call([])
+            self.assertEqual(1, mock_run.call_count)
```
For such a big change, updating one unittest doesn't sound enough. Can you please add more tests as per https://github.com/letsencrypt/letsencrypt/issues/787#issuecomment-143837334?

pde replied:
I agree that more tests are advisable, but I'm not going to block these bugfixes on writing them; opening #867 instead.

kuba replied:
CLI has been bugfixed a lot of times, often introducing regressions. I'm afraid that this will happen again, if we just don't provide tests :( In fact, how am I meant to test this PR? Test everything manually myself? It's really hard to verify just by looking at the code.

kuba replied:
backref to regression from this PR: https://github.com/letsencrypt/letsencrypt/pull/863#discussion_r40856338

pde replied:
Actually, since there may be a couple of bugs still in this PR, I'll write test cases now.

kuba replied:
SGTM! :+1:

kuba commented on letsencrypt/cli.py:
```diff
@@ -723,10 +730,18 @@ def create_parser(plugins, args):
+    "config_changes" : config_changes,
+    "plugins"        : plugins_cmd
+}
+HELP_TOPICS = (["all", "security", "paths", "automation", "testing", "apache", "nginx"]
```
Do you have to list all plugin names here manually? If so, then any new registered plugins (even those not controlled by us) will not be automatically registered here. In fact, you're missing `manual` and it's `--test-mode`.

kuba replied:
To make the example even more concrete... https://github.com/e00E/lets-encrypt-icecast/blob/master/icecast.py#L137-L141 plugin help will not be visible due to hard coding the topics here.

kuba replied:
In fact, this is one of the regressions mentioned in https://github.com/letsencrypt/letsencrypt/pull/863#discussion_r40856138. :(

pde replied:
I don't think they need to be listed.

kuba replied:
Then why are they [nginx,apache] listed? :O Why don't we add tests and find out? :D

pde replied:
I've removed them :).  Working on adding those tests, though.

bmw commented:
I also want to mention that I don't think #787 should be closed (at least not without opening additional issues). The following problems mentioned there are not fixed:

* `--apache`, `--standalone`, and `--nginx` do not work
* `letsencrypt [SUBCOMMANDS] [options] [domains]` is incorrect. The CLI does not accept domains as the final option on the command line (see #865).

kuba commented:
#819/#837 also doesn't seem to be fixed due to the `--help $plugin` regression mentioned above

pde commented:
@kuba you mention a regression; can you give me an example input that has regressed?

pde commented:
Landing!

kuba commented:
@pde, did anyone re-review it?

kuba commented on letsencrypt/cli.py:
```diff
@@ -724,77 +731,79 @@ def create_parser(plugins, args):
+        if name == "plugins":
+            func = plugins_cmd
+        else:
+            func = eval(name) # pylint: disable=eval-used
```
Pylint's `eval-used` exists for purpose. It definitely should NOT be used this way :(

kuba commented on letsencrypt/display/enhancements.py:
```diff
@@ -11,7 +11,7 @@
 logger = logging.getLogger(__name__)
 
 # Define a helper function to avoid verbose code
-util = zope.component.getUtility  # pylint: disable=invalid-name
```
Again, may I please ask you to refrain updating code that is irrelevant to the PR?

kuba commented on letsencrypt/tests/cli_test.py:
```diff
@@ -42,12 +43,54 @@ def _call(self, args):
+            self.assertRaises(SystemExit, self._call_stdout, ['-h'])
+            out = output.getvalue()
+            from letsencrypt import cli
+            self.assertTrue(cli.USAGE in out)
```
there is too much stuff happening in this test, whenever you do `output.truncate(0)` you should create a new unit test

kuba commented on letsencrypt/tests/cli_test.py:
```diff
@@ -42,12 +43,54 @@ def _call(self, args):
+        with mock.patch('letsencrypt.cli.sys.stderr') as stderr:
+            with mock.patch('letsencrypt.cli.client') as client:
+                ret = cli.main(args)
+        return ret, None, stderr, client
```
According to Travis and coverage testing this line is never hit! https://travis-ci.org/letsencrypt/letsencrypt/jobs/83033097#L585

pde replied:
Yes, it's currently only called within assertRaises().

kuba replied:
We have a policy of deleting any dead code, such as this line.
